app/src/main/java/io/apicurio/registry/rules/RulesServiceImpl.java

```diff
@@ -21,24 +22,74 @@
-                }
-                break;
-            case SYNTAX_VALIDATION:
-                break;
-            case SEMANTIC_VALIDATION:
-                break;
+    
+    @Inject
+    RuleExecutorFactory factory;
+    
+    /**
+     * @see io.apicurio.registry.rules.RulesService#applyRules(java.lang.String, io.apicurio.registry.types.ArtifactType, java.lang.String, io.apicurio.registry.rules.RuleApplicationType)
+     */
+    @Override
+    public void applyRules(String artifactId, ArtifactType artifactType, String artifactContent,
+            RuleApplicationType ruleApplicationType) throws RuleViolationException {
+        boolean useGlobalRules = false;
+        @SuppressWarnings("unchecked")
+        List<String> rules = Collections.EMPTY_LIST;
+        if (ruleApplicationType == RuleApplicationType.update) {
+            rules = storage.getArtifactRules(artifactId);
+        }
+        if (rules.isEmpty()) {
+            rules = storage.getGlobalRules();
+            useGlobalRules = true;
+        }
+        if (rules.isEmpty()) {
+            return;
+        }
+        String currentArtifactContent = null;
+        if (ruleApplicationType == RuleApplicationType.update) {
+            StoredArtifact currentArtifact = storage.getArtifact(artifactId);
+            currentArtifactContent = currentArtifact.content;
+        }
+        for (String ruleName : rules) {
+            RuleType ruleType = RuleType.valueOf(ruleName);
+            RuleConfigurationDto configurationDto = useGlobalRules ? 
+                    storage.getGlobalRule(ruleName) : storage.getArtifactRule(artifactId, ruleName);
+            RuleExecutor executor = factory.createExecutor(ruleType);
+            RuleContext context = new RuleContext(artifactId, artifactType, configurationDto.getConfiguration(), 
+                    currentArtifactContent, artifactContent);
+            executor.execute(context);
         }
     }
-
-    public boolean isCompatible(ArtifactType type, CompatibilityLevel level, String artifactId, String content) {
-        StoredArtifact artifact = storage.getArtifact(artifactId); // check against the last one
-        ArtifactTypeAdapter adapter = ArtifactTypeAdapterFactory.toAdapter(type);
-        return adapter.isCompatibleWith(level.name(), Collections.singletonList(artifact.content), content);
+    
+    /**
+     * @see io.apicurio.registry.rules.RulesService#applyRule(java.lang.String, io.apicurio.registry.types.ArtifactType, java.lang.String, io.apicurio.registry.types.RuleType, java.lang.String, io.apicurio.registry.rules.RuleApplicationType)
+     */
+    @Override
+    public void applyRule(String artifactId, ArtifactType artifactType, String artifactContent,
+            RuleType ruleType, String ruleConfiguration, RuleApplicationType ruleApplicationType)
+            throws RuleViolationException {
+        String currentArtifactContent = null;
```

Comment:
This code is duplicated here https://github.com/Apicurio/apicurio-registry/pull/54/files#diff-f971af92a4148e2ca0536172340654feR48